airbyte-integrations/connectors/source-amazon-ads/source_amazon_ads/streams/report_streams/report_streams.py

```diff
@@ -0,0 +1,345 @@
+    """
+
+
+class ReportStream(BasicAmazonAdsStream, ABC):
+    """
+    Common base class for report streams
+    """
+
+    primary_key = None
+    CHECK_INTERVAL_SECONDS = 30
+    REPORT_WAIT_TIMEOUT = timedelta(hours=1)
+    # Format used to specify metric generation date over Amazon Ads API.
+    REPORT_DATE_FORMAT = "%Y%m%d"
+    cursor_field = "reportDate"
+
+    def __init__(self, config: AmazonAdsConfig, profiles: List[Profile], authenticator: Oauth2Authenticator):
+        self._authenticator = authenticator
+        self._session = requests.Session()
+        self._model = self._generate_model()
+        # Set start date from config file, should be in UTC timezone.
+        self._start_date = pendulum.parse(config.start_date).set(tz="UTC") if config.start_date else None
+        super().__init__(config, profiles)
+
+    @property
+    def model(self) -> CatalogModel:
+        return self._model
+
+    def read_records(
+        self,
+        sync_mode: SyncMode,
+        cursor_field: List[str] = None,
+        stream_slice: Mapping[str, Any] = None,
+        stream_state: Mapping[str, Any] = None,
+    ) -> Iterable[Mapping[str, Any]]:
+        """
+        This is base method of CDK Stream class for getting metrics report. It
+        collects metrics for all profiles and record types. Amazon Ads metric
+        generation works in async way: First we need to initiate creating report
+        for specific profile/record type/date and then constantly check for report
+        generation status - when it will have "SUCCESS" status then download the
+        report and parse result.
+        """
+
+        if not stream_slice:
+            # In case of stream_slices method returned [None] object and
+            # stream_slice is None.
+            # This may occure if we have abnormal stream state with start_date
+            # parameter pointing to the future. In this case we dont need to
+            # take any action and just return.
+            return
+        report_date = stream_slice[self.cursor_field]
+        report_infos = self._init_reports(report_date)
+        logger.info(f"Waiting for {len(report_infos)} report(s) to be generated")
+        # According to Amazon Ads API docs metric generation takes maximum 15
+        # minutes. But in case reports wont be generated we dont want this stream to
+        # hung forever. Store timepoint when report generation has started to
+        # check if it takes to long to break a loop.
+        start_time_point = datetime.now()
+        while report_infos and datetime.now() <= start_time_point + self.REPORT_WAIT_TIMEOUT:
+            completed_reports = []
+            logger.info(f"Checking report status, {len(report_infos)} report(s) remained")
+            for report_info in report_infos:
+                report_status, download_url = self._check_status(report_info)
+                if report_status == Status.FAILURE:
```

Comment:
there is one dangerous behavior here that can cause data loss. Let's say all reports failed for some reason. Then we would just log "generation failed", then continue running the connection incrementally sync after sync, as if everything worked successfully, correct? And we would have a success in the UI? that would be a silent failure

Can we instead raise an exception if a report failed or if requesting a report failed?

Reply:
Make sense, updated